[1 - 3단계 방탈출 사용자 예약] 제프리(홍성호) 미션 제출합니다.

AppleMint98 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?

> https://prolog.techcourse.co.kr/studylogs/4304
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
API 구현과 DB 스키마 및 데이터 문제해결에 집중하느라, 객체지향 생활체조 요구사항을 제대로 신경쓰지 못했습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

안녕하세요 찰리! 제프리입니다 반갑습니다 😄 
궁금한점은 코드에 코멘트로 달아놓겠습니다! 잘부탁드려요 🙇 
[방탈출 사용자 예약 변경사항 커밋 목록](https://github.com/woowacourse/spring-roomescape-member/pull/239/files/0c7122a6d5c9c3dd58d1ba2384cac1e8253429d9..9774d92c50b4f90d7b5346b988aa5d1cfdbd0e00)

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

AppleMint98 committed:
Initial Commit

phk1128 committed:
chore: 벨리데이션 및 mockito 의존성 추가

phk1128 committed:
refactor: 통합 테스트 성공하게끔 리팩토링

phk1128 committed:
refactor: 이미 예약된 시간을 삭제할 떄 예외가 발생하게끔 수정

AppleMint98 committed:
refactor: 지나간 날짜 또는 시간으로 예약 생성시 예외 발생하도록 수정

AppleMint98 committed:
fix: 입력 데이터 유효한 시간 수정

AppleMint98 committed:
feat: 테마 도메인 초안 구현

phk1128 committed:
feat: 테마를 포함한 예약 생성 기능 구현

phk1128 committed:
fix: 예약 생성 기능 오류 수정

AppleMint98 committed:
feat: 인기테마 조회기능 구현

AppleMint98 committed:
fix: 테스트 컴파일 에러 수정

AppleMint98 committed:
fix: MissionStepTest 수정

phk1128 committed:
fix: 컨트롤러 및 레파지토리 테스트 깨지는 부분 수정

phk1128 committed:
test: 테마 컨트롤러 및 레파지토리 테스트 코드 작성

AppleMint98 committed:
feat: ExceptionHandler 적용

AppleMint98 committed:
refactor: 불필요한 생성자 제거

AppleMint98 committed:
refactor: 중복 prefix 통합

AppleMint98 committed:
refactor: 매직넘버 제거

applemint98 commented on src/main/java/roomescape/config/WebConfig.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.web.servlet.config.annotation.ViewControllerRegistry;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+@Configuration
+public class WebConfig implements WebMvcConfigurer {
+
+    @Override
+    public void addViewControllers(ViewControllerRegistry registry) {
+        registry.addViewController("/reservation").setViewName("reservation");
+        registry.addViewController("/admin").setViewName("admin/index");
+    }
+}
```
현재 `/reservation` , `/admin` 과 같은 기본 페이지는 WebConfig의 `addViewControllers()` 를 통해 등록하고, `/admin/reservation` , `/admin/time` 처럼 추가 로직이 필요한 페이지는 AdminPageController에서 처리하고 있습니다.

이와 같은 구조에서 HTML을 반환하는 API를 별도 컨트롤러로 분리하는 것이 좋을지, 아니면 `addViewControllers()` 를 계속 활용하는 것이 좋을지 고민 중입니다.

또한, WebMvcConfigurer를 구현한 설정 클래스는 어떤 경우에 사용하는 것이 적절한지도 궁금합니다.

Gomding replied:
말씀하신대로 추가 로직이 없고 단순 정적 리소스 반환이라면 `WebMvcConfigurer` 를 사용하는것도 방법입니다 :)

> WebMvcConfigurer를 구현한 설정 클래스는 어떤 경우에 사용하는 것이 적절한지도 궁금합니다.

`WebMvcConfigurer` 너무 다양한 활용방향이 있어서 전부 답변드리기는 어려울 것 같아요 😢 
설정이 다양하기에 어떤 경우에 사용하기보다 어떤 기능들이 있는지 학습하시는게 좋아요!
어떤 기능이 있는지 알고있으면 제프리가 필요한 경우에 자유롭게 사용하면 됩니다 :)

applemint98 replied:
감사합니다!!
덕분에 [`WebMvcConfigurer`](https://www.geeksforgeeks.org/spring-mvc-webmvcconfigure/) 에 대해 더욱 찾아봤습니다 😄 

공식문서에는
> Defines callback methods to customize the Java-based configuration for Spring MVC enabled via @EnableWebMvc.

라고 적혀있는데, 

`스프링이 기본적으로 제공하는 설정들(인터셉터, 메시지 컨버터, 뷰 리졸버 등)을 개발자가 원하는 방향으로 바꿀 수 있도록 콜백 메서드를 제공한다` 라고 이해했습니다. 해당 인터페이스는 `addViewControllers` `addInterceptors` `extendMessageConverters` `configureViewResolvers` 등의 메서드를 정의하고 있더라구요!

해당 인터셉터, 컨버터 등의 개념들은 아직 잘 모르지만, 기본적으로 세 가지 종류의 접두사 (`add` , `extend` , `configure`) 만 사용하는게 눈에 걸려 추가적으로 찾아보았는데, 

`add` : 기본 설정이 없는 빈들에 대하여 새로운 빈을 추가함
`extend` : 수정자를 통해 기존의 설정을 대신하여 등록함
`configure` : 기존의 설정을 이용하며 추가로 설정을 확장함

라는 차이점이 있다는 것까지 알았습니다.

해당 [링크](https://mangkyu.tistory.com/176)를 참고했습니다! 😃

applemint98 commented on src/main/java/roomescape/repository/RoomescapeThemeRepositoryImpl.java:
```diff
@@ -0,0 +1,83 @@
+
+    @Override
+    public List<ReservationTheme> findWeeklyThemeOrderByCountDesc() {
+        String sql = """
+                SELECT th.id, th.name, th.description, th.thumbnail, COUNT(*) AS reservation_count
+                FROM reservation r
+                JOIN reservation_theme th ON r.theme_id = th.id
+                WHERE PARSEDATETIME(r.date, 'yyyy-MM-dd') BETWEEN DATEADD('DAY', -7, CURRENT_DATE) AND DATEADD('DAY', -1, CURRENT_DATE)
+                GROUP BY th.id, th.name, th.description, th.thumbnail
+                ORDER BY reservation_count DESC
+                LIMIT 10; 
+                """;
```
현재 스키마에서는 시간 데이터를 문자열(VARCHAR) 타입으로 관리하고 있습니다.

하지만 이 경우, `최근 1주일 내` 와 같은 시간 기반 비즈니스 로직을 구현할 때 쿼리가 복잡해지고 성능 저하가 우려되는 문제가 있습니다.

이런 상황에서 시간 데이터를 문자열로 관리하는 것이 적절한 경우는 언제인지,
또는 어떤 기준으로 DATETIME과 같은 타입을 사용하는 것이 좋은지 궁금합니다.

Gomding replied:
시간 유형이 있는데 `VARCHAR` 를 사용하는 이유가 궁금해요!

> 이런 상황에서 시간 데이터를 문자열로 관리하는 것이 적절한 경우는 언제인지,
> 또는 어떤 기준으로 DATETIME과 같은 타입을 사용하는 것이 좋은지 궁금합니다.

시간 데이터를 문자열로 관리하는 것에 장점이 떠오르지 않네요...

applemint98 replied:
미션에서 기본적으로 제공하는 스키마에서 VARCHAR를 사용해서 별다른 의심없이 사용하고 있었습니다!
저도 VARCHAR는 오히려 입력값에 추가적인 검증이 필요하다는 점에서 별로라고 생각합니다.

DATE로 바꿔볼게요 ㅎㅎ 감사합니다!! 👍

applemint98 commented on src/main/java/roomescape/repository/RoomescapeRepositoryImpl.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Override
+    public Reservation save(final Reservation reservation) {
+        String sql = "insert into reservation (name, date, time_id, theme_id) values (?, ?, ?, ?)";
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        template.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(sql, new String[]{"id"});
+            ps.setString(1, reservation.getName());
+            ps.setString(2, reservation.getDate().toString());
+            ps.setLong(3, reservation.getTime().getId());
+            ps.setLong(4, reservation.getTheme().getId());
+            return ps;
+        }, keyHolder);
+
+        long id = keyHolder.getKey().longValue();
+        return reservation.toEntity(id);
+    }
```
SimpleJdbcInsert와 NamedParameterJdbcTemplate을 사용하도록 변경해서, 
다음 리뷰요청에 추가적으로 구현하여 제출하겠습니다!

Gomding commented on README.md:
```diff
@@ -0,0 +1,25 @@
+## 방탈출 예약 관리
```
이번 미션에 대한 내용 중 일부가 README 에 없는것 같네요 :)
요구사항을 추가해보면 좋겠어요!
(요구사항을 먼저 정리하는것은 중요합니다! ㅎㅎ)

applemint98 replied:
앗 추가하도록 하겠습니다!! 😃

Gomding commented on src/main/java/roomescape/controller/ReservationThemeController.java:
```diff
@@ -0,0 +1,54 @@
+        this.roomescapeService = roomescapeService;
+    }
+
+    @GetMapping()
```
`()` 는 생략해도 되겠군요 :)

```suggestion
    @GetMapping
```

Gomding commented on src/main/java/roomescape/controller/ReservationThemeController.java:
```diff
@@ -0,0 +1,54 @@
+        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+    }
+
+    @ExceptionHandler(value = IllegalArgumentException.class)
+    public ResponseEntity<String> handleIllegalArgumentException() {
+        return ResponseEntity.badRequest().build();
+    }
```
응답에 에러 메세지는 전달하지 않아도 괜찮을까요 :)
API 를 호출한 클라이언트가 어떤 에러 상황인지 알 수 없을것같아요!

applemint98 replied:
에러 메세지도 전달하도록 수정하겠습니다 🙇

Gomding commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,48 @@
+        return ResponseEntity.noContent().build();
+    }
+
+    @ExceptionHandler(value = IllegalArgumentException.class)
+    public ResponseEntity<String> handleIllegalArgumentException() {
+        return ResponseEntity.notFound().build();
+    }
```
`IllegalArgumentException` 에 대해서 무조건 notFound 를 응답해줘도 될지 궁금해요!
특히 `IllegalArgumentException`은 java 표준 예외인데 다양한 상황에 사용될 것 같네요 :)

applemint98 replied:
맞습니다!! `IllegalArgumentException` 을 사용하기에는 다양한 상황을 고려해서, 커스텀 예외를 사용하도록 만들어볼게요!
감사합니다 😄

Gomding commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,66 @@
+    private final ReservationTime time;
+    private final ReservationTheme theme;
+
+    private Reservation(final long id, final String name, final LocalDate date, final ReservationTime time,
+                        final ReservationTheme theme) {
+        this.id = id;
```
Reservation 객체 생성 시 인자가 올바르게 넘어왔는지 검사하지 않아도 될까요?
Reservation 객체의 규칙은 없는지 궁금합니다!

applemint98 replied:
초기에는 생각하지 못했는데, 이름 글자 수 제한 등이 필요할 것 같아요! 
감사합니다!

Gomding commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,66 @@
+        this.theme = theme;
+    }
+
+    public Reservation(final long id, final String name, final String date, final ReservationTime time,
+                       final ReservationTheme theme) {
+        this.id = id;
+        this.name = name;
+        this.date = LocalDate.parse(date);
+        this.time = time;
+        this.theme = theme;
+    }
+
+    public Reservation(final String name, final LocalDate date, final ReservationTime time,
+                       final ReservationTheme theme) {
+        this.name = name;
+        this.date = date;
+        this.time = time;
+        this.theme = theme;
+    }
```
주 생성자를 호출하도록 변경해보면 좋겠어요!

Gomding commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,66 @@
+        this.theme = theme;
+    }
+
+    public Reservation toEntity(long id) {
+        return new Reservation(id, name, date, time, theme);
+    }
```
어떤 의미의 메서드인지 궁금해요 🤔

applemint98 replied:
초기에는 id없이 생성한 Reservation에 id를 새롭게 부여해주는 메서드였으나,
현재 id를 직접 인자로 받아 생성하는 생성자를 만들면서 필요없을 것 같아요!

추가로,` id를 직접 인자로 받아 생성하는 생성자` 대신` id는 세터로 할당` 하는 방법으로 변경하였는데, 
해당 방법에서도 `toEntity()` 라는 네이밍이 id를 할당하는 의도를 드러내기에 너무 추상적이라고 생각해서,
`assignId()` 로 이름을 변경하였습니다!

Gomding replied:
적절한 이름을 부여해주는것은 좋아요!
assignId 메서드를 반영하면서 객체의 불변이 깨진것 같은데
불변이 깨진것에 대해서 제프리는 어떻게 생각하시나요! 😃 
저는 내부에서 새로운 객체를 만들어 반환하는것은 좋았다고 생각해요

applemint98 replied:
id이외의 필드를 final로 설정했기 때문에 주요 상태에 대해서는 불변성이 지켜진다고 생각했는데, 
이왕이면 새로운 객체를 만들어 반환하는게 더욱 확실하게 불변성을 지킬 수 있어 좋아보입니다!
수정하겠습니다 😄

Gomding commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,66 @@
+
+import java.time.LocalDate;
+
+public class Reservation {
```
equals & hashCode 를 재정의 해보면 좋겠어요!

Gomding commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,66 @@
+
+public class Reservation {
+
+    private Long id;
```
primitive 타입인 long도 있는데
id를 Long 타입으로 하신 배경이 궁금해요!

applemint98 replied:
초기에는 long으로 구현을 했었는데, id 값을 할당하지 않으면 기본값인 0으로 할당되어서 바꿨습니다.

id가 할당되지 않으면 데이터를 삽입할 때 에러가 나오는 것을 기대했는데, 
id가 0인 데이터로 정상적으로 db에 들어가는 사이드 이펙트가 발생해서 변경하게 되었습니다!

Gomding commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,66 @@
+    private final ReservationTime time;
+    private final ReservationTheme theme;
+
+    private Reservation(final long id, final String name, final LocalDate date, final ReservationTime time,
```
id 필드와 타입이 동일하지 않군요 😢

applemint98 replied:
수정하겠습니다!! 감사합니다!

Gomding reviewed (changes requested):
안녕하세요 제프리!
방탈출 사용자 예약 미션 함께하게된 찰리입니다 :)
미션 잘 진행해주셨네요 👍 👍 
몇가지 코멘트 남겼으니 확인부탁드려요~

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요!

AppleMint98 committed:
refactor(repository): NamedParameterJdbcTemplate & SimpleJdbcInsert 사용하도록 변경

AppleMint98 committed:
test(repository): 반환값 변경에 따른 컴파일 에러 수정

AppleMint98 committed:
feat(ControllerExceptionHandler): 잘못된 삭제요청 에러처리 ControllerAdvice 구현

AppleMint98 committed:
refactor(schema): 시간 타입 VARCHAR -> DATETIME 변경

AppleMint98 committed:
refactor(controller): 미사용 괄호 삭제

AppleMint98 committed:
feat(exception): 커스텀 예외 구현

DataNotFoundException: 해당되는 id의 데이터를 찾지 못하는 경우
DuplicateReservationException: 중복된 예약을 생성하는 경우
PastReservationTimeException: 현재시간 이전의 예약을 생성하는 경우

AppleMint98 committed:
refactor(Reservation): 불필요한 생성자 제거

AppleMint98 committed:
refactor(ReservationTheme): 불필요한 생성자 제거

AppleMint98 committed:
refactor(ReservationTime): 불필요한 생성자 제거

AppleMint98 committed:
build: 롬복 애노테이션 추가

AppleMint98 committed:
feat(domain): EqualsAndHashCode 재정의

AppleMint98 committed:
refactor(domain): Getter 메서드 롬복 사용 변경

AppleMint98 committed:
refactor(dto): Request DTO 빈 데이터 검증 추가

AppleMint98 committed:
feat(domain): 이름 글자 수 제한 추가

AppleMint98 committed:
refactor(RoomescapeThemeRepositoryImpl): 인기테마 날짜기준 동적 쿼리로 변경

AppleMint98 committed:
refactor(RoomescapeService): 메서드 분리

AppleMint98 committed:
docs(README): 요구사항 업데이트

Gomding commented on src/main/java/roomescape/repository/RoomescapeTimeRepositoryImpl.java:
```diff
@@ -0,0 +1,75 @@
+            String sql = "delete from reservation_time where id = :id";
+            SqlParameterSource param = new MapSqlParameterSource("id", id);
+            return template.update(sql, param) == SUCCESS_COUNT;
+        } catch (DataIntegrityViolationException e) {
+            throw new DeletionNotAllowedException("예약 시간을 지울 수 없습니다.");
+        }
```
해당 메세지를 받고 클라이언트가 다음 행동을 취할 수 있는지 고민해보면 좋겠어요 😃 
예외 메세지는 호출한 쪽에 주는 피드백이라고 생각하시면 좋아요!
친절한 피드백은 에러 상황을 알 수 있고 다음 행동을 어떻게 취해야할지 알 수 있게 해주는거라 생각해요!

- https://hyeon9mak.github.io/why-should-throw-detail-error-message/
- https://hyeon9mak.github.io/writing-helpful-error-messages/

+) 현재는 `DataIntegrityViolationException` 을 처리하고있기 때문에 자세한 처리가 불가능할 것 같은데요
어떻게 개선할 수 있을지 고민해보면 좋겠어요!!

applemint98 replied:
감사합니다!
첨부해주신 링크를 보니, 확실히 자세한 예외 메세지를 주는 것이 더욱 좋아보입니다!
현재는 안된다 라고만 제시하기 때문에 해결방안을 알기 어렵네요.. 수정하겠습니다 😄 

본래 의도는 외래키 제약 조건을 위배하는 SQL의 경우 에러를 던지는 것이었는데, 
현재 [`DataIntegreityViolationException`](https://www.baeldung.com/spring-dataintegrityviolationexception)은 총 세가지 원인이 있어서, 분리가 필요할 것 같아요!

에러메세지를 구체화시키기에는 에는 `DataIntegreityViolationException` 보다는 `ConstraintViolationException` 으로 잡는것이 올바를 것 같아, 해당 에러로 바꿔보겠습니다!

applemint98 replied:
해당 주제에 대해 더 생각해보았는데, Service 레이어에서 
```java
if (reservationRepository.existsByThemeId(id)) {
    throw new DeletionNotAllowedException("해당 테마에 예약이 있어 삭제할 수 없습니다.");
}
```
이런식으로 변경하는 것이 더 좋을 것 같아요. 

그 이유는, 
초기에 데이터 삽입 및 삭제의 선후관계는 외래키 제약과 같은 DB 차원의 문제라고 생각했는데, 
그렇게 되면 예외처리가 DB에 의존하게 된다고 생각합니다.
만약 DB를 MongoDB 등 RDBMS 이외의 DB로 변경하거나, 스키마에서 외래키 제약조건을 삭제시킨다고 가정할 때
예외를 서비스레이어세서 관리할 수 있으면 코드 변경에 더욱 견고해질 것 같아요.

Gomding replied:
말씀대로 `DataIntegrityViolationException` 은 DB 의 에러로 볼 수 있는데 
해당 예외로 비즈니스 흐름에 대한 규칙을 처리하기는 한계가 있다고 생각해요 👍

Gomding commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,43 @@
+
+    @EqualsAndHashCode.Include
+    private Long id;
+    @Size(max = 5, message = "이름은 다섯글자 이내로 입력해 주세요.")
+    private final String name;
```
`@Size` 를 사용하신 이유가 궁금해요!
이름을 다섯글자 초과하도록 Reservation 객체를 생성하면 현재 검사를 하고있을까요?
테스트도 작성해보면 좋겠어요 :)
레벨1을 떠올려보죠!!

applemint98 replied:
초기에 `@Size` 를 사용한 이유는 , [`SpringBootAnnotations`](https://medium.com/@himani.prasad016/validations-in-spring-boot-e9948aa6286b) 에서 문자열 크기를 지정하는 어노테이션을 봤기 때문입니다!

현재 코드의 문제점은, `@Valid` 를 호출한 컨트롤러의 
```java
    @PostMapping
    public ResponseEntity<ReservationResponse> reservationAdd(@Valid @RequestBody ReservationRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(roomescapeService.addReservation(request));
    }
```

코드에서, 
```java
public record ReservationRequest(
        @NotBlank
        String name,
        @DateTimeFormat(pattern = "yyyy-MM-dd")
        LocalDate date,
        @NotNull
        Long themeId,
        @NotNull
        Long timeId) {
}
```

해당 어노테이션인 `@NotBlack` 만 적용되고, `@Size`  는 직접적으로 적용되지 않는 문제가 있었네요!
검증 로직을 도메인 내부로 넣어보겠습니다!! 감사합니다! 😄

Gomding commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,43 @@
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
+public class Reservation {
+
+    @EqualsAndHashCode.Include
+    private Long id;
```
아래 테스트를 성공하는게 맞을까요?

```java
        ReservationTime time = new ReservationTime("10:30");
        ReservationTime time2 = new ReservationTime("10:31");
        assertThat(time).isEqualTo(time2);
```

객체 둘이 의미적으로 다르고 id 가 null일때 동등 비교를 하면 결과가 어떻게 나와야하는지도 고민해볼 수 있어요 😃

applemint98 replied:
와 생각하지도 못했습니다.. 감사합니다!
id로 EqualsAndHashCode를 재정의하는것에 더불어 
null 이 왔을 때의 동등 비교를 생각하지 못했네요! 

id가 null 인 상황은, 아직 PK값을 부여받지 못한 상황이므로 해당 테스트는 false가 나와야한다고 생각합니다.
같은 객체라고 판단하는 조건은, 동일한 PK를 가지고 있는 경우에 한해서만이라고 생각합니다.
지금 방식의 재정의는 id가 null인 두 객체에 대해 `isEqualTo()` 가 true를 반환합니다. 

여기에대해서 주변 크루들에게 물어본 결과 두가지 해결방법을 찾았습니다.
1. id 포함한 클래스의 모든 필드로 eqauls를 재정의한다.
2. 재정의한 equals에서 id=null일경우 false를 반환하는 로직을 추가. `this.getId() != null`
```java
@Override
public boolean equals(final Object o) {
    if (o == null || getClass() != o.getClass()) {
        return false;
    }
    ReservationTheme that = (ReservationTheme) o;
    return this.getId() != null && Objects.equals(id, that.id);
}
```

해당 해결방법에 대해서는 제 기준에서 찝찝한 면이 있었습니다.
1. id가 식별자인데, id외의 필드를 추가해서 equals & hashcode 를 재정의하는게 별로라고 생각했습니다. 식별자의 존재의의가 희석된다고 생각했기 때문입니다.
2. IDE에서 정해준 equals & hashcode 외의 로직을 개발자가 직접 설정하기 싫었습니다. 동등성을 보장하기 위한 equals() 메서드를 재정의하기 위해 지켜야 할 규칙이 많은 것으로 알고있는데, 새로 도메인을 추가할때마다 equals&hashcode의 코드를 직접 건드려야한다는게 별로라고 생각했습니다. 대부분의 개발자는 IDE가 생성한 메서드 그대로 사용할 것이기 때문에, 추후 휴먼에러가 나타날 수 있다고 생각했습니다.


이런 고민들을 하다가, 근본적인 원인이 해당 도메인이 반쪽짜리 엔티티이기 때문이라는 결론에 도달했습니다.
`new Reservation()` 을 통해 생성하면 단순 VO처럼 생성되고, 추후 `assignId()` 를통해 id를 할당해줌으로써 Entity가 된다고 생각합니다. 
이런 이유때문에 고유 식별자인 PK가 식별자의 기능을 제대로 수행하지 못한다고 생각했습니다.
왜 이렇지?를 생각해봤을때, AutoIncrement를 통해 PK를 할당받기 위해서는 DB를 거쳐야 합니다. DB를 거쳐 가장 최근에 생성된 값을 알아야, 해당 값보다 1이 증가된 새로운 값을 id로 할당해 줄 수 있기 때문입니다.

그래서 DB를 거치지 않고 식별자를 추가해 줄 수 있는 id를 uuid로 변경하기를 고민중입니다.
uuid를 통해 객체생성시 ID를 할당해준다면 식별자의 기능이 정상적으로 동작한다고 생각합니다.

이런 이유들로 id값을 uuid로 바꿀까를 고민했는데, 찰리의 의견이 궁금합니다! 😄

applemint98 replied:
추가로 서비스에서 레포지토리사이에도 전달용 객체가 있으면 해결될 것 같은데, 

ex)
```java
public class ReservationTime {

    private Long id;
    private final LocalTime startAt;
}

public class ReservationTimeDto {

    private final LocalTime startAt;
}
```

오버엔지니어링인것같아 좋아보이지는 않게 느껴집니다. 😢

Gomding replied:
uuid 로 했을 경우의 단점도 고민은 해봐야할거에요!
AutoIncrement 일때의 장점도 있을 것이구요 :)

> IDE에서 정해준 equals & hashcode 외의 로직을 개발자가 직접 설정하기 싫었습니다. 동등성을 보장하기 위한 equals() 메서드를 재정의하기 위해 지켜야 할 규칙이 많은 것으로 알고있는데, 새로 도메인을 추가할때마다 equals&hashcode의 코드를 직접 건드려야한다는게 별로라고 생각했습니다. 대부분의 개발자는 IDE가 생성한 메서드 그대로 사용할 것이기 때문에, 추후 휴먼에러가 나타날 수 있다고 생각했습니다.

보통 식별자로 동등하다고 판단하는 객체를 `Entity` 라고 부릅니다
모든 필드값을 기준으로 동등하다고 판단하는것을 값 객체(VO) 라고 부를 수 있구요
현재 필요한건 Entity 는 항상 식별자(id)로 동등성을 판단할거고 `equals&hashCode` 는 항상 동일한 규칙을 가질 것 같네요 :)
레벨1 블랙잭과 장기미션을 떠올리면서 추상클래스를 활용해보는것은 어떻게 생각하시나요?

```
abstract class EntityObject {
      ~~
}

class Reservation extends EntityObject {
    ~~
}
```

Gomding replied:
> 추가로 서비스에서 레포지토리사이에도 전달용 객체가 있으면 해결될 것 같은데,

어떤 부분을 해결하기위해 사용하는것인지 이해하지 못했어요 😢

Gomding commented on src/main/java/roomescape/repository/RoomescapeRepositoryImpl.java:
```diff
@@ -0,0 +1,120 @@
+        return Boolean.TRUE.equals(template.queryForObject(sql, param, Boolean.class));
+    }
+
+    private RowMapper<Reservation> reservationRowMapper() {
```
RowMapper 는 매번 생성해줘야하는 객체일까요 :)
한번 생성해놓고 재사용할 수 있으면 좋겠어요~

applemint98 replied:
생각해보지 못했어요!!

감사합니다!! 😄

Gomding commented on src/main/java/roomescape/repository/RoomescapeRepositoryImpl.java:
```diff
@@ -0,0 +1,120 @@
+                    rs.getDate("date").toLocalDate(),
+                    reservationTime,
+                    reservationTheme
+            ).assignId(rs.getLong("reservation_id"));
```
id 를 받는 생성자를 Reservation, ReservationTheme, ReservationTime 에 만들어놓으면 별도로 assignId 를 호출하지 않아도 될 것 같네요 🤔 

혹시 주 생성자를 사용하라는 피드백을 적용하면서 이렇게 변경한 걸까요?
`주 생성자와 부 생성자` 라는것을 학습해보면 좋겠어요!

- https://tecoble.techcourse.co.kr/post/2021-05-17-constructor/

applemint98 replied:
먼저 `assignId()` 메서드를 만들면서, 
굳이 id를 받는 생성자를 만들지 않아도 `assignId()` 를 통해 id를 추후 할당해줄 수 있을거라고 생각해 지웠습니다!

다시생각해보니 불변성의 측면에서 안좋을것같습니다.
`주생성자` 와 `부생성자` 를  활용하면, 제가이해한바로는, 

```java
public ReservationTime(final Long id, final LocalTime startAt) {
    this.id = id;
    this.startAt = startAt;
}

public ReservationTime(final LocalTime startAt) {
    this(null, startAt);
}

public ReservationTime assignId(final Long id) {
    return new ReservationTime(id, startAt);
}
```

이런식으로 작성하면 코드가독성도 좋게 생성자를 만들 수 있을 것 같아요!

변경하겠습니다~! 👍

Gomding commented on src/main/java/roomescape/domain/ReservationTime.java:
```diff
@@ -0,0 +1,31 @@
+        this.startAt = startAt;
+    }
+
+    public ReservationTime(final String input) {
+        this(LocalTime.parse(input));
+    }
```
잘못된 format 으로 된 문자열을 넘기면 어떤 예외가 발생할까요?
클라이언트는 어떤 메세지를 받고있을지 궁금해요!

applemint98 replied:
`DateTimeParseException` 이 발생하네요!!
추가하겠습니다!! 감사합니다!

Gomding commented on build.gradle:
```diff
@@ -27,8 +27,13 @@ dependencies {
-
     testImplementation 'io.rest-assured:spring-mock-mvc'
+
+    // Lombok
+    compileOnly 'org.projectlombok:lombok'
+    annotationProcessor 'org.projectlombok:lombok'
+    testCompileOnly 'org.projectlombok:lombok'
+    testAnnotationProcessor 'org.projectlombok:lombok'
```
lombok 사용없이도 구현이 가능했을텐데  
lombok 을 사용하신 이유가 궁금해요 :)

applemint98 replied:
게터 등 보일러플레이트 코드를 제거하기 위해 사용했습니다!

그동안 롬복은 무조건적으로 사용해야하는, 좋은 것이라는 인식이 있었는데, 
남겨주신 코멘트 덕분에 롬복의 단점에 대해 공부해볼 수 있었습니다.

가장 큰 문제점은 
타입이 같다면, 필드의 순서가 바뀔 때 다른 필드여도 컴파일에러없이 실행이 된다는 점인 것 같습니다. 
[참고자료](https://kwonnam.pe.kr/wiki/java/lombok/pitfall) 를 확인했을 때, 
`AllArgsConstructor`, `RequiredArgsConstructor` 의 생성자 어노테이션의 경우는 사용을 금지하는 것이 좋다고 나오네요!

제가 사용한 `EqualsAndHashCode` 에 대해서도, 변경가능한 객체에 대해서만 사용하는 게 좋다고 합니다.
더욱 불변으로 만들어야 할 이유가 생기는것 같아요 ㅎㅎ

감사합니다!!!! 👍

Gomding replied:
말씀대로 보일러 플레이트 코드 작성에 대한 시간을 아껴주는데
그만큼 학습이 필요하고 공유해주신 링크처럼 조심해야하는 포인트가 있습니다
알수없는 동작을 할 때 디버깅을 해야하는 책임도 존재하구요

`AllArgsConstructor, RequiredArgsConstructor` 도 분명 편리하고 좋은 애너테이션이지만
말씀해주신것처럼 필드 순서 변경에 치명적인 단점이 있다고 볼 수 있죠 :)

해당 부분을 인지하고 있으니 보일러 플레이트 코드를 작성하는 시간을 아끼자는 취지로 사용은 하지만 규칙을 강하게 세우는 방법이 있을 것이구요
그게아니면 아예 사용하지 않는것도 방법이라고 생각해요!

제프리가 함께할 팀에서 잘 논의하여 방향을 세워보면 좋겠네요 😃

Gomding commented on src/main/java/roomescape/exception/handler/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,34 @@
+import roomescape.exception.exception.PastReservationTimeException;
+
+@ControllerAdvice
+public class ControllerExceptionHandler {
```
전체적으로 커스텀 예외를 만들어서 잘 핸들링 해주신것 같네요 👍 👍
제프리가 예상하지 못한 예외가 발생하면 클라이언트에게 정제되지 않은 예외메세지가 노출될 것 같은데
해당 부분은 어떻게 처리하면 좋을까요?

applemint98 replied:
앗 최상위의 Exception을 받는 ExceptionHandler도 만들어야겠어요!!
감사합니다 찰리~! 😄

Gomding commented on src/main/java/roomescape/exception/handler/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,34 @@
+@ControllerAdvice
+public class ControllerExceptionHandler {
+
+    @ExceptionHandler(DataNotFoundException.class)
+    public ResponseEntity<String> handleDataNotFoundException(final DataNotFoundException ex) {
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ex.getMessage());
```
예외 응답에 문자열 외에 표현이 더 필요한 상황이 오면 현재는 body가 단순 문자열이라 값 추가가 힘들 것 같아요!
예외 응답 객체를 구현해보는건 어떻게 생각하시나요? 😃

applemint98 replied:
생각하지 못했던 내용이었어요! 해보겠습니다!

Gomding replied:
약간의 TMI 를 하자면
응답 Json 의 형태에서 최상위는 항상 Object 가 되는것이 좋습니다
(String or Array 도 아닌)
구조 변경이 제일 적고 변화에 쉽게 대응할 수 있어서요 :)
응답에 추가/변경이 있을 때 백엔드 개발자와 클라이언트 개발자 모두 최소한의 공수로 수정할 수 있습니다

최상위가 String 인 구조
- 응답에 필드를 추가하고 싶으면 최상위를 Object 로 변경해야함
```
"ㅁㄴㅇㄹ"
```


최상위가 Array 인 구조
- 배열 바깥에 필드를 추가하고 싶으면 최상위를 Object 로 변경해야함
```
[
    {
      "age":12
    },
    {
      "age":13
    }
]
```

최상위가 Object 인 구조
- 배열 바깥에 필드가 필요하면 구조 변경없이 간단하게 추가
```
{
  "something": [
    {
      "age":12
    },
    {
      "age":13
    }
  ],
  "addField": "쉽게 추가할 수 있음"
}
```

Gomding commented on src/main/java/roomescape/service/RoomescapeService.java:
```diff
@@ -0,0 +1,133 @@
+import roomescape.repository.RoomescapeTimeRepository;
+
+@Service
+public class RoomescapeService {
```
`RoomescapeService` 가 많은 책임을 가지고 있는것으로 보여요 🤔 
Service 를 분리해보는것은 어떻게 생각하시나요?

Gomding commented on src/main/java/roomescape/dto/ReservationTimeResponse.java:
```diff
@@ -0,0 +1,11 @@
+import java.time.LocalTime;
+import roomescape.domain.ReservationTime;
+
+public record ReservationTimeResponse(long timeId, LocalTime startAt) {
```
응답객체에 시간의 format 을 결정해주지 않아도 괜찮을까요!

applemint98 replied:
놓친 부분이네요! 추가하겠습니다 😃

Gomding commented on src/test/java/roomescape/service/RoomescapeServiceTest.java:
```diff
@@ -0,0 +1,71 @@
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+class RoomescapeServiceTest {
```
`RoomescapeService` 에 행위가 많은데 테스트코드는 일부만 작성됐군요 🧐
제프리의 테스트 작성 기준이 궁금해요!

applemint98 replied:
누락된 테스트들은 보충하겠습니다 😢 

Service같은 경우는, 현재는 예외처리 위주로 잡고있습니다.
현재 메서드들 대부분이 단순 Repository의 데이터를 DTO로 변환해주기만 하기 때문에, 그런 메서드들에 한해서는 성공 테스트가 필요하지 않다고 생각했습니다. 해당 테스트들을 RepositoryTest에서 수행하고 있다고 생각했거든요.

현재처럼 Service의 메서드가 단순히 Repository에서 얻어온 데이터를 변환하는 용도여도, 테스트를 작성해야 할까요??
```java
public List<ReservationThemeResponse> findReservationThemes() {
    List<ReservationTheme> reservationThemes = themeRepository.findAll();
    return reservationThemes.stream().map(ReservationThemeResponse::of).toList();
}
```

추가로 서비스관련한 테스트를 작성하는 영역(?), 어느 메서드까지 테스트를 해야하는가? 에 대해서 기준을 잘 세우지 못했습니다. 😢 
물론 전부다 하는게 제일 좋긴 하겠지만, 작성할수록 레포지토리 영역의 테스트와 동일한 테스트라는 느낌을 지울 수 없었습니다. 
찰리의 기준을 여쭤봐도 될까요..? 🙇

Gomding replied:
말씀대로 Service 에서 단순 Repository 를 호출하는 메서드에 대한 테스트이고 Repository 테스트가 존재한다면
생략할 수는 있을 것 같아요!
핵심은 Service 에서의 테스트로 검증하고자 하는 부분과 Repository 에서 테스트로 검증하는 부분이 완전 일치하고있는가? 일 것 같네요 :)

테스트 작성 영역에 대해서는 제프리가 테스트로 커버가 필요하다고 느끼는 부분을 모두 작성하시면 됩니다 :)
모든 메서드에 대해 테스트를 작성을 해봐도 되고 
필요하다고 생각한 일부만 작성을 했더니 어떤 케이스를 못잡아내더라 하면서 테스트를 보강하고 다시 기준을 고민해볼 수 있겠죠!

Gomding reviewed (changes requested):
안녕하세요 제프리!
다음단계 넘어가기전에 몇가지 더 의견나눠보면 좋겠어요 😃 

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요!!

AppleMint98 committed:
refactor(repository): 예외 메세지 수정

AppleMint98 committed:
feat(RoomescapeRepository): 시간ID, 테마ID로 예약 존재 검사 로직 추가

AppleMint98 committed:
refactor(RoomescapeService): 외래키 제약 조건 위배 검증 서비스레이어로 이동

AppleMint98 committed:
test(RoomescapeServiceTest): 외래키 제약조건 위배 검증 테스트 작성

AppleMint98 committed:
fix(Reservation): 이름 검증 로직 수정

기존 이름 글자 수 검증로직을 수정하고, 한국어만 입력가능하도록 검증을 추가합니다.

AppleMint98 committed:
fix(ReservationTheme): 테마명 검증 로직 수정

AppleMint98 committed:
test(RoomescapeServiceTest): 예약자 성함 수정

AppleMint98 committed:
test(database): test profile 설정

기존 프로덕션 코드의 data.sql을 통한 데이터 삽입이 테스트코드에 영향을 미쳐, 분리하기 위해 별도의 프로파일을 설정하였습니다.

AppleMint98 committed:
fix(domain): 생성자 추가, 객체 불변 반환 메서드 수정

id를 추가로 받는 주 생성자를 추가하였고,
기존 불변적으로 반환하지 않던 assignId를 새로운 객체를 생성하여 반환하도록 수정합니다

AppleMint98 committed:
refactor(domain): rowMapper 필드 변경

AppleMint98 committed:
feat(ReservationTime): 잘못된 시간 문자열 포맷 검증

AppleMint98 committed:
feat(ControllerExceptionHandler): 최상위 예외 ExceptionHandler 추가

AppleMint98 committed:
feat(ControllerExceptionHandler): @DateTimeFormat 문자열 형식 예외 ExceptionHandler 추가

AppleMint98 committed:
feat(ErrorResponse): 예외 응답 객체 구현

AppleMint98 committed:
refactor(RoomescapeService): 서비스 분리

AppleMint98 committed:
test(RoomescapeServiceTest): 서비스 분리

AppleMint98 committed:
test(ReservationServiceTest): 존재하지 않는 예약시간, 예약테마 테스트 추가

AppleMint98 committed:
refactor(response): Date, Time 포맷 결정

Gomding commented on README.md:
```diff
@@ -0,0 +1,235 @@
+## 방탈출 예약 관리
```
지금처럼 각 API 문서를 정리해보는것도 좋아요

레벨1과 동일하게 구현해야하는 요구사항에 대해 정리해보시면 좋겠어요 :)
예를들면 아래와 같이요

- [ ] 예약 시간을 삭제할 수 있다
  - [ ] 예약 시간은 `HH:mm` 형태를 가진다 (ex. 10:00)
  - [ ] 존재하지 않는 예약 시간은 삭제할 수 없다
  - [ ] 해당 시간으로 예약이 존재하면 예약은 삭제할 수 없다

Gomding commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,62 @@
+                       final ReservationTheme theme) {
+        validateNameKoreanWords(name);
+        validateNameSize(name);
+        this.id = id;
+        this.name = name;
+        this.date = date;
+        this.time = time;
+        this.theme = theme;
```
각 필드에 대해 null 인지 검사는 하지않아도 될까요? :)
DTO 에서 검사를 진행하고 있어서 도메인 객체에서는 검사하지 않았다면
도메인 객체가 항상 DTO의 검증된 값으로부터 생성된다고 확신할 수 있는지 고민해볼 수 있어요!

Gomding commented on src/main/java/roomescape/exception/handler/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,46 @@
+import roomescape.exception.exception.DuplicateReservationException;
+import roomescape.exception.exception.PastReservationTimeException;
+
+@ControllerAdvice
```
`@RestControllerAdvice` 를 고려해볼 수 있겠네요 :)

Gomding reviewed (approved):
안녕하세요 제프리!
깊게 고민하신 흔적이 보이네요 👍 👍 👍 
이번 단계는 이만 머지하겠습니다 :)
추가로 남긴 코멘트는 다음 단계 진행하면서 함께 고민해주세요!

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요
